vaadin-time-picker-flow-parent/vaadin-time-picker-flow/src/main/resources/META-INF/resources/frontend/timepickerConnector.js

```diff
@@ -133,112 +75,85 @@
-              let cleanedTimeString = localeTimeString;
-              if (localeTimeString.endsWith(amString)) {
-                cleanedTimeString = localeTimeString.replace(' ' + amString, '');
-              } else if (localeTimeString.endsWith(pmString)) {
-                cleanedTimeString = localeTimeString.replace(' ' + pmString, '');
-              }
-              if (milliseconds) {
-                let millisecondsString = milliseconds < 10 ? '0' : '';
-                millisecondsString += milliseconds < 100 ? '0' : '';
-                millisecondsString += milliseconds;
-                cleanedTimeString += '.' + millisecondsString;
-              } else {
-                cleanedTimeString += '.000';
-              }
-              if (localeTimeString.endsWith(amString)) {
-                cleanedTimeString = cleanedTimeString + ' ' + amString;
-              } else if (localeTimeString.endsWith(pmString)) {
-                cleanedTimeString = cleanedTimeString + ' ' + pmString;
-              }
-              return cleanedTimeString;
-            }
-            return localeTimeString;
-          };
-
           let cachedTimeString;
           let cachedTimeObject;
 
           timepicker.i18n = {
             formatTime: tryCatchWrapper(function (timeObject) {
-              if (timeObject) {
-                let timeToBeFormatted = new Date();
-                timeToBeFormatted.setHours(timeObject.hours);
-                timeToBeFormatted.setMinutes(timeObject.minutes);
-                timeToBeFormatted.setSeconds(timeObject.seconds !== undefined ? timeObject.seconds : 0);
-                let localeTimeString = timeToBeFormatted.toLocaleTimeString(locale, getTimeFormatOptions());
-                // milliseconds not part of the time format API
-                localeTimeString = formatMilliseconds(localeTimeString, timeObject.milliseconds);
-                return localeTimeString;
+              if (!timeObject) return;
+
+              const timeToBeFormatted = new Date();
+              timeToBeFormatted.setHours(timeObject.hours);
+              timeToBeFormatted.setMinutes(timeObject.minutes);
+              timeToBeFormatted.setSeconds(timeObject.seconds !== undefined ? timeObject.seconds : 0);
+
+              // the web component expects the correct granularity used for the time string,
+              // thus need to format the time object in correct granularity by passing the format options
+              let localeTimeString = timeToBeFormatted.toLocaleTimeString(locale, {
+                hour: 'numeric',
+                minute: 'numeric',
+                second: includeSeconds() ? 'numeric' : undefined
+              });
+
+              // milliseconds not part of the time format API
+              if (includeMilliSeconds()) {
+                localeTimeString = formatMilliseconds(localeTimeString, timeObject.milliseconds, amString, pmString);
```

Comment:
**note:** Previously, the `includeMilliSeconds` check was done inside `formatMilliseconds`.